### PR TITLE
8325517: Shenandoah: Reduce unnecessary includes from shenandoahControlThread.cpp

### PR DESCRIPTION
**Pull request by earthling-amzn**
The control thread used to run much more of the cycle directly. This code was all factored out into different classes, but many of the vestigial headers remained. Removing these improves compilation times and makes maintenance easier.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325517](https://bugs.openjdk.org/browse/JDK-8325517): Shenandoah: Reduce unnecessary includes from shenandoahControlThread.cpp (**Task** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role) ⚠️ Review applies to [fd24ad00](https://git.openjdk.org/jdk/pull/17778/files/fd24ad00e3a3a07d74aff76f0de1708201221ed0)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**) ⚠️ Review applies to [fd24ad00](https://git.openjdk.org/jdk/pull/17778/files/fd24ad00e3a3a07d74aff76f0de1708201221ed0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17778/head:pull/17778` \
`$ git checkout pull/17778`

Update a local copy of the PR: \
`$ git checkout pull/17778` \
`$ git pull https://git.openjdk.org/jdk.git pull/17778/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17778`

View PR using the GUI difftool: \
`$ git pr show -t 17778`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17778.diff">https://git.openjdk.org/jdk/pull/17778.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17778#issuecomment-1934720654)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahUtils.hpp [hunk 1]
```diff
@@ -42,6 +42,13 @@
 
 class GCTimer;
 
+
+enum StringDedupMode {
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUtils.hpp:46]: I think the only use is in `ShenandoahMark`, right? Therefore, this declaration should go to `shenandoahMark.hpp`, so we do not have to include `shenandoahUtils.hpp` there?

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 2]
```diff
@@ -34,17 +34,12 @@
...
 #include "gc/shenandoah/shenandoahRootProcessor.inline.hpp"
 #include "gc/shenandoah/shenandoahUtils.hpp"
 #include "gc/shenandoah/shenandoahVMOperations.hpp"
-#include "gc/shenandoah/shenandoahWorkerPolicy.hpp"
 #include "gc/shenandoah/heuristics/shenandoahHeuristics.hpp"
-#include "memory/iterator.hpp"
 #include "memory/metaspaceUtils.hpp"
 #include "memory/metaspaceStats.hpp"
-#include "memory/universe.hpp"
-#include "runtime/atomic.hpp"
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:47]: `Atomic` is directly used, so the include should be left here.

**Review by shipilev — COMMENTED** [review]
Looks generally okay. The usual rule of thumb is to see if any symbols from the include are actually used in the compilation unit. If used, keep the explicit header. Check this? At least one case is obvious from reading the code, see below.

Note that build might still succeed without the explicit header due to transitive header dependencies, but it is fragile and would yield build failures later.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp (hunk 2 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:47]: Yes, sorry. The branch I cherry-picked from no longer uses `Atomic` directly. I'll put it back.

**Comment by earthling-amzn** [comment]
I only removed a couple of headers that were obviously not needed directly in the control thread (I think there are a couple more I can trim). I didn't try to find and declare all the direct/first-level dependencies (`log.hpp`, for example).

**Review by kdnilsen — APPROVED** [review]

**Review by ysramakrishna — APPROVED** [review]
Looks good modulo Aleksey's suggestion on checking use of header file's content in compilation unit rather than relying on serendipitous (and fragile) transitive dependencies.

**Review by shipilev — APPROVED** [review]

**Pull request closed without merging**